File: python-ai/services/AIIngestService.py
"""
AI Ingest Service
Handles document ingestion and processing
"""

import os
from typing import List, Dict, Optional
from utils.fileUtils import (
    scan_directory, read_file_content, get_file_metadata,
    calculate_file_hash, get_project_name, detect_project_type
)
from utils.textUtils import chunk_text, clean_text, extract_summary, extract_keywords
from utils.logger import get_logger

logger = get_logger(__name__)


class AIIngestService:
    """Service for ingesting and processing files"""

    def __init__(self):
        self.processing_queue = []
        self.is_processing = False

    def ingest_project(self, project_path: str) -> Dict:
        """Ingest project files"""
        logger.info(f"Starting project ingestion: {project_path}")

        try:
            project_name = get_project_name(project_path)
            project_type = detect_project_type(project_path)
            files = scan_directory(project_path)

            logger.info(f"Found {len(files)} files to ingest")

            results = {
                'project_name': project_name,
                'project_path': project_path,
                'project_type': project_type,
                'files_processed': 0,
                'files_skipped': 0,
                'errors': [],
            }

            for file_path in files:
                try:
                    self.ingest_file(file_path, project_name, project_path, project_type)
                    results['files_processed'] += 1
                except Exception as e:
                    logger.error(f"Error ingesting file: {file_path}, {str(e)}")
                    results['files_skipped'] += 1
                    results['errors'].append({'file': file_path, 'error': str(e)})

            logger.info(f"Project ingestion completed: {results}")
            return results
        except Exception as e:
            logger.error(f"Error ingesting project: {project_path}, {str(e)}")
            raise

    def ingest_file(self, file_path: str, project_name: str, project_path: str, project_type: str) -> Dict:
        """Ingest single file"""
        logger.debug(f"Ingesting file: {file_path}")

        # Read file content
        content = read_file_content(file_path)
        if not content:
            raise Exception("Could not read file content")

        # Get file metadata
        metadata = get_file_metadata(file_path)
        if not metadata:
            raise Exception("Could not extract file metadata")

        # Calculate file hash
        file_hash = calculate_file_hash(file_path)

        # Clean and process content
        cleaned_content = clean_text(content)
        summary = extract_summary(cleaned_content)
        keywords = extract_keywords(cleaned_content)

        # Create chunks
        chunks = chunk_text(cleaned_content)
        chunk_data = [
            {
                'chunk_index': idx,
                'content': chunk,
                'chunk_id': f"{file_hash}-{idx}",
            }
            for idx, chunk in enumerate(chunks)
        ]

        # Create document data
        document_data = {
            'filename': metadata['filename'],
            'filepath': file_path,
            'project_name': project_name,
            'project_path': project_path,
            'project_type': project_type,
            'language': metadata['language'],
            'file_type': metadata['file_type'],
            'file_size': metadata['file_size'],
            'file_hash': file_hash,
            'content': cleaned_content,
            'summary': summary,
            'keywords': keywords,
            'chunks': chunk_data,
            'total_chunks': len(chunks),
            'processed': False,
            'embeddings_generated': False,
            'last_modified_at': metadata['modified_at'],
        }

        logger.debug(f"Ingested file: {file_path}, chunks: {len(chunks)}")
        return document_data

    def get_unprocessed_documents(self, limit: int = 100) -> List[Dict]:
        """Get unprocessed documents"""
        # This will be implemented with MongoDB integration
        return []

    def mark_as_processed(self, document_id: str):
        """Mark document as processed"""
        # This will be implemented with MongoDB integration
        pass

    def mark_embeddings_generated(self, document_id: str):
        """Mark embeddings as generated"""
        # This will be implemented with MongoDB integration
        pass

    def get_documents_by_project(self, project_name: str) -> List[Dict]:
        """Get documents by project"""
        # This will be implemented with MongoDB integration
        return []

    def get_statistics(self) -> Dict:
        """Get ingestion statistics"""
        # This will be implemented with MongoDB integration
        return {
            'total_documents': 0,
            'processed_documents': 0,
            'documents_with_embeddings': 0,
            'total_chunks': 0,
            'by_language': {},
            'by_file_type': {},
            'by_project': {},
        }

    def delete_project_documents(self, project_name: str) -> Dict:
        """Delete documents by project"""
        # This will be implemented with MongoDB integration
        logger.info(f"Deleted project documents: {project_name}")
        return {'deleted_count': 0}

    def search_documents(self, query: str, limit: int = 10) -> List[Dict]:
        """Search documents"""
        # This will be implemented with MongoDB integration
        return []

File: backend/ai/services/AIIngestService.js
/**
 * AI Ingest Service
 * Handles document ingestion and processing
 */

const fs = require('fs');
const path = require('path');
const AIDocument = require('../models/AIDocument');
const { scanDirectory, readFileContent, getFileMetadata, calculateFileHash, getProjectName, detectProjectType } = require('../utils/fileUtils');
const { chunkText, cleanText, extractSummary, extractKeywords } = require('../utils/textUtils');
const AILogger = require('../utils/logger');
const AI_CONFIG = require('../config/aiConfig');

const logger = new AILogger('IngestService');

class AIIngestService {
  constructor() {
    this.processingQueue = [];
    this.isProcessing = false;
  }



  async ingestProject(projectPath) {
    logger.info('Starting project ingestion', { projectPath });

    try {
      const projectName = getProjectName(projectPath);
      const projectType = detectProjectType(projectPath);
      const files = scanDirectory(projectPath);

      logger.info('Found files to ingest', { projectPath, fileCount: files.length });

      const results = {
        projectName,
        projectPath,
        projectType,
        filesProcessed: 0,
        filesSkipped: 0,
        errors: [],
      };

      for (const filePath of files) {
        try {
          await this.ingestFile(filePath, projectName, projectPath, projectType);
          results.filesProcessed++;
        } catch (err) {
          logger.error('Error ingesting file', { filePath, error: err.message });
          results.filesSkipped++;
          results.errors.push({ filePath, error: err.message });
        }
      }

      logger.info('Project ingestion completed', results);
      return results;
    } catch (err) {
      logger.error('Error ingesting project', { projectPath, error: err.message });
      throw err;
    }
  }

  /**
   * Ingest single file
   */
  async ingestFile(filePath, projectName, projectPath, projectType) {
    logger.debug('Ingesting file', { filePath });

    // Read file content
    const content = readFileContent(filePath);
    if (!content) {
      throw new Error('Could not read file content');
    }

    // Get file metadata
    const metadata = getFileMetadata(filePath);
    if (!metadata) {
      throw new Error('Could not extract file metadata');
    }

    // Calculate file hash
    const fileHash = calculateFileHash(filePath);

    // Check if file already exists and hasn't changed
    const existingDoc = await AIDocument.findOne({ filepath: filePath });
    if (existingDoc && existingDoc.fileHash === fileHash) {
      logger.debug('File unchanged, skipping', { filePath });
      return existingDoc;
    }

    // Clean and process content
    const cleanedContent = cleanText(content);
    const summary = extractSummary(cleanedContent);
    const keywords = extractKeywords(cleanedContent);

    // Create chunks
    const chunks = chunkText(cleanedContent);
    const chunkData = chunks.map((chunk, index) => ({
      chunkIndex: index,
      content: chunk,
      chunkId: `${fileHash}-${index}`,
    }));

    // Create or update document
    const documentData = {
      filename: metadata.filename,
      filepath: filePath,
      projectName,
      projectPath,
      projectType,
      language: metadata.language,
      fileType: metadata.fileType,
      fileSize: metadata.fileSize,
      fileHash,
      content: cleanedContent,
      summary,
      keywords,
      chunks: chunkData,
      totalChunks: chunks.length,
      processed: false,
      embeddingsGenerated: false,
      lastModifiedAt: metadata.modifiedAt,
    };

    let document;
    if (existingDoc) {
      document = await AIDocument.findByIdAndUpdate(existingDoc._id, documentData, { new: true });
      logger.debug('Updated existing document', { filePath });
    } else {
      document = await AIDocument.create(documentData);
      logger.debug('Created new document', { filePath });
    }

    return document;
  }

  /**
   * Get unprocessed documents
   */
  async getUnprocessedDocuments(limit = 100) {
    return AIDocument.find({ processed: false }).limit(limit);
  }

  /**
   * Mark document as processed
   */
  async markAsProcessed(documentId) {
    return AIDocument.findByIdAndUpdate(
      documentId,
      { 
        processed: true,
        lastProcessedAt: new Date(),
      },
      { new: true }
    );
  }

  /**
   * Mark document embeddings as generated
   */
  async markEmbeddingsGenerated(documentId) {
    return AIDocument.findByIdAndUpdate(
      documentId,
      { embeddingsGenerated: true },
      { new: true }
    );
  }

  /**
   * Get documents by project
   */
  async getDocumentsByProject(projectName) {
    return AIDocument.find({ projectName });
  }

  /**
   * Get document statistics
   */
  async getStatistics() {
    const stats = {
      totalDocuments: await AIDocument.countDocuments(),
      processedDocuments: await AIDocument.countDocuments({ processed: true }),
      documentsWithEmbeddings: await AIDocument.countDocuments({ embeddingsGenerated: true }),
      totalChunks: 0,
      byLanguage: {},
      byFileType: {},
      byProject: {},
    };

    const documents = await AIDocument.find();
    for (const doc of documents) {
      stats.totalChunks += doc.totalChunks;
      stats.byLanguage[doc.language] = (stats.byLanguage[doc.language] || 0) + 1;
      stats.byFileType[doc.fileType] = (stats.byFileType[doc.fileType] || 0) + 1;
      stats.byProject[doc.projectName] = (stats.byProject[doc.projectName] || 0) + 1;
    }

    return stats;
  }

  /**
   * Delete documents by project
   */
  async deleteProjectDocuments(projectName) {
    const result = await AIDocument.deleteMany({ projectName });
    logger.info('Deleted project documents', { projectName, deletedCount: result.deletedCount });
    return result;
  }

  /**
   * Search documents
   */
  async searchDocuments(query, limit = 10) {
    return AIDocument.find(
      { $text: { $search: query } },
      { score: { $meta: 'textScore' } }
    )
    .sort({ score: { $meta: 'textScore' } })
    .limit(limit);
  }
}

module.exports = AIIngestService;
